### DISCUSSION TITLE
ChampSim full documentation

### DISCUSSION THREAD

**Discussion by SaffatZabin-17** [ChampSim/ChampSim#500] [Q&A] [answered]
Hi. I am implementing a new cache prefetching technique using ChampSim as part of my undergraduate thesis. However, after going through the documentation in the champsim.github.io website, I do not yet have the clear idea of the construction of ChampSim to get working on my project. 

By documentation, I mean the detailed architecture of the system that ChampSim is simulating, the cache architecture, how are the instructions fed to ChampSim, how the cache operates those instructions, how the L1C, L2C and LLC is implemented, where the prefetches are stored etc. Overall, a detailed documentation on the inner workings of ChampSim. 

Is there any such type of documentation? TIA

**Comment by maccoymerrell** [comment, marked as answer]
The paper (here https://arxiv.org/pdf/2210.14324) has some additional information on the overall structure.

The system being simulated is defined largely by your config file. The base config models a single-core super-scalar processor, with a 3-level cache (using LRU replacement) and a memory controller to DRAM. Their internal parameters are set through the config file, along with some policies such as replacement for the cache.

To understand ChampSim's implementation, it would be best to familiarize yourself with how these systems are designed generally (through existing external sources) which will make ChampSim's simulation easier to follow. 
If you are familiar with C++ programming, I would also advise you to locate the operate() function within each operable and understand the procedure each part of ChampSim goes through in each cycle.

I hope this helps.

**Comment by ngober** [comment]
Maccoy's answer is correct and practical, in particular that he notes that nearly everything in ChampSim's internal structure is configurable (if you're willing to wrestle the configuration file). This means that any documentation must be sufficiently general as to not answer your question. See [this section of the docs](https://champsim.github.io/ChampSim/master/Creating-a-configuration-file.html#heterogeneous-systems) for an example of what I mean.

I'll add a philosophical point that hints at an engineering decision: If something is documented, people will rely on it and it becomes difficult to change. What that means is that you should only document things that you expect the users of your software to interact with routinely. In ChampSim, we provide the module system as the preferred way of interacting with the simulator, and we [document it here](https://champsim.github.io/ChampSim/master/Modules.html). 

We could document somewhere that the upstrems of a cache are always handled in order by CPU index, which means that packets that arrive in the same cycle will always be ordered in that way. I can understand how that might be useful microarchitectural information to have. But, then, we would not have the flexibility we need to fix #454. That's just an example. The engineering decision is then this: focus the documentation on the expected interaction, but make the underlying system comprehensible enough to the person who wants to dig into it. If you venture that way, congratulations! We have a lot of fun working on this stuff and we hope you'll fix something while you're down there.
